File: _wpf/wpf-business-programmer/3-wpf-business-programmer-m3-exercise-files/Samples/VB/WPFTemplateApp/WPFTemplateApp/ViewModels/EmployeeViewModel.vb
Imports System.Collections.ObjectModel
Imports System.Windows

Public Class EmployeeViewModel
  Inherits ViewModelBase
#Region "Constructor"
  Public Sub New()
    ' Make sure GetEmployees has great exception handling!
    DataCollection = _DataManager.GetEmployees()
  End Sub
#End Region

#Region "Private Variables"
  Private _DataManager As New EmployeeManager()
#End Region

#Region "DataCollection Property"
  Public Property DataCollection() As ObservableCollection(Of Employee)
    Get
      Return DirectCast(GetValue(DataCollectionProperty), ObservableCollection(Of Employee))
    End Get
    Set(ByVal value As ObservableCollection(Of Employee))
      SetValue(DataCollectionProperty, value)
    End Set
  End Property

  Public Shared ReadOnly DataCollectionProperty As DependencyProperty = DependencyProperty.Register("DataCollection", GetType(ObservableCollection(Of Employee)), GetType(EmployeeViewModel), New UIPropertyMetadata(Nothing))
#End Region

#Region "DetailData Property"
  Public Property DetailData() As Employee
    Get
      Return DirectCast(GetValue(DetailDataProperty), Employee)
    End Get
    Set(ByVal value As Employee)
      SetValue(DetailDataProperty, value)
    End Set
  End Property

  Public Shared ReadOnly DetailDataProperty As DependencyProperty = DependencyProperty.Register("DetailData", GetType(Employee), GetType(EmployeeViewModel), New UIPropertyMetadata(Nothing))
#End Region

#Region "AddRecord Method"
  Public Sub AddRecord()
    SetViewStateMode(UIStateMode.Add)

    ' Create Empty Object for UI to Display
    DetailData = New Employee()
  End Sub
#End Region

#Region "CancelEdit Method"
  Public Sub CancelEdit()
    SetViewStateMode(UIStateMode.Cancel)

    ' TODO: Write Code to Undo Here
  End Sub
#End Region

#Region "SaveData Method"
  Public Sub SaveData()
    If IsAddMode Then
      _DataManager.Insert(DetailData)
      DataCollection.Add(DetailData)
      SetViewStateMode(UIStateMode.Normal)
    Else
      _DataManager.Update(DetailData)
      SetViewStateMode(UIStateMode.Normal)
    End If
  End Sub
#End Region

#Region "DeleteData Method"
  Public Sub DeleteData()
    _DataManager.Delete(DetailData)
    DataCollection.Remove(DetailData)
    SetViewStateMode(UIStateMode.Normal)
  End Sub
#End Region
End Class

File: _wpf/wpf-business-programmer/3-wpf-business-programmer-m3-exercise-files/Samples-2012/CSharp/WPFTemplateApp/WPFTemplateApp/ViewModels/EmployeeViewModel.cs
using System.Collections.ObjectModel;
using System.Windows;

namespace WPFTemplateApp
{
  public class EmployeeViewModel : ViewModelBase
  {
    #region Constructor
    public EmployeeViewModel()
    {
      // Make sure GetEmployees has great exception handling!
      DataCollection = _DataManager.GetEmployees();
    }
    #endregion

    #region Private Variables
    EmployeeManager _DataManager = new EmployeeManager();
    #endregion

    #region DataCollection Property
    public ObservableCollection<Employee> DataCollection
    {
      get { return (ObservableCollection<Employee>)GetValue(DataCollectionProperty); }
      set { SetValue(DataCollectionProperty, value); }
    }

    public static readonly DependencyProperty DataCollectionProperty =
        DependencyProperty.Register("DataCollection", typeof(ObservableCollection<Employee>), typeof(EmployeeViewModel), new UIPropertyMetadata(null));
    #endregion
    
    #region DetailData Property
    public Employee DetailData
    {
      get { return (Employee)GetValue(DetailDataProperty); }
      set { SetValue(DetailDataProperty, value); }
    }

    public static readonly DependencyProperty DetailDataProperty =
        DependencyProperty.Register("DetailData", typeof(Employee), typeof(EmployeeViewModel), new UIPropertyMetadata(null));
    #endregion

    #region AddRecord Method
    public void AddRecord()
    {
      SetViewStateMode(UIStateMode.Add);

      // Create Empty Object for UI to Display
      DetailData = new Employee();
    }
    #endregion

    #region CancelEdit Method
    public void CancelEdit()
    {
      SetViewStateMode(UIStateMode.Cancel);

      // TODO: Write Code to Undo Here
    }
    #endregion

    #region SaveData Method
    public void SaveData()
    {
      if (IsAddMode)
      {
        _DataManager.Insert(DetailData);
        DataCollection.Add(DetailData);
        SetViewStateMode(UIStateMode.Normal);
      }
      else
      {
        _DataManager.Update(DetailData);
        SetViewStateMode(UIStateMode.Normal);
      }
    }
    #endregion

    #region DeleteData Method
    public void DeleteData()
    {
      _DataManager.Delete(DetailData);
      DataCollection.Remove(DetailData);
      SetViewStateMode(UIStateMode.Normal);
    }
    #endregion
  }
}
